### PR TITLE
Patch: Undefined file name

### PR DESCRIPTION
**Pull request by mitchellst**
I was getting a weird error where the file would upload to google cloud with the right name, but I'd get `undefined` returned to ghost.

e.g., cloud storage would have `2018/04/myfile.jpg`, but in my post editor it would show as: `2018/04undefined`.

Variable hoisting FTW.

### CODE REVIEW

**Comment by Uysim** [comment]
@gcochard  I think you should merge this as soon as possible

**Comment by gcochard** [comment]
_I am not an admin on this repo so I can't merge this PR_

With that said, this looks good to me other than the leading comments on index.js, which should be dropped before anyone merges.

File: index.js
Comment (gcochard) [inline index.js:5]: Drop these comments.

**Comment by Uysim** [comment]
I think @thombuchi can merge that

**Comment by thombuchi** [comment]
I’m without my computer until thursday, then i will be able to test the pull request and merge it.

**Comment by Uysim** [comment]
@thombuchi how is it going with your testing?

**Comment by ollaollu** [comment]
Hi, will this ever get merged? It is still an issue @thombuchi

**Review by lcotonea — CHANGES_REQUESTED** [review]
The target url to google storage seems wrong. And the usage of targetDir introduce a new bug into the url path.
Here a new version of the patch:

```
'use strict';

var storage     = require('@google-cloud/storage'),
    BaseStore   = require('ghost-storage-base'),
    Promise     = require('bluebird'),
    options     = {};

class GStore extends BaseStore {
    constructor(config = {}){
        super(config);
        options = config;

        var gcs = storage({
            projectId: options.projectId,
            keyFilename: options.key
        });
        this.bucket = gcs.bucket(options.bucket);
        this.assetDomain = options.assetDomain || `storage.googleapis.com/${options.bucket}`;
        // only set insecure from config if assetDomain is set
        if(options.hasOwnProperty('assetDomain')){
            this.insecure = options.insecure;
        }
        // default max-age is 3600 for GCS, override to something more useful
        this.maxAge = options.maxAge || 2678400;
    }

    save(image) {
        if (!options) return Promise.reject('google cloud storage is not configured');

        var targetDir = this.getTargetDir(),
        googleStoragePath = `http${this.insecure?'':'s'}://${this.assetDomain}/`,
        targetFilename;

        return new Promise((resolve, reject) => {
            this.getUniqueFileName(image, targetDir).then(tf => {
                targetFilename = tf;
                var opts = {
                    destination: targetFilename,
                    metadata: {
                        cacheControl: `public, max-age=${this.maxAge}`
                    },
                    public: true
                };
                return this.bucket.upload(image.path, opts);
            }).then(function (data) {
                return resolve(googleStoragePath + targetFilename);
            }).catch(function (e) {
                return reject(e);
            });
        });
    }

    // middleware for serving the files
    serve() {
        // a no-op, these are absolute URLs
        return function (req, res, next) { next(); };
    }

    exists (filename) {
        return new Promise((resolve, reject) => {
            this.bucket.file(filename).exists().then(function(data){
                return resolve(data[0]);
            });
        });
    }

    read (filename) {
      var rs = this.bucket.file(filename).createReadStream(), contents = '';
      return new Promise(function (resolve, reject) {
        rs.on('error', function(err){
          return reject(err);
        });
        rs.on('data', function(data){
          contents += data;
        });
        rs.on('end', function(){
          return resolve(content);
        });
      });
    }

    delete (filename) {
        return this.bucket.file(filename).delete();
    }
}

module.exports = GStore;

```

**Comment by dfloresdev** [comment]
the solution of @mitchellst it's good, but not return the complete URL, missing the name of bucket, the solution is in the variable **store** in the file of configuration
````bash
"storage": {
    "active": "gcloud",
    "gcloud": {
      "projectId": "projectId",
      "key": "key.json",
      "bucket": "name_of_bucket",
      "insecure": false,
      "assetDomain": "storage.cloud.google.com/name_of_bucket",
      "maxAge": "time"
    }
}
````

**Comment by JonnieDoe** [comment]
@thombuchi Is this going to be merged?

**Comment by thombuchi** [comment]
Merged

**Pull request merged**
